google/cloud/pubsublite/internal/resumable_async_streaming_read_write_rpc.h

```diff
@@ -0,0 +1,352 @@
+          if (!start_future.get()) {
+            {
+              std::lock_guard<std::mutex> g{this->mu_};
+              this->in_retry_loop_ = true;
+            }
+
+            this->AttemptRetry(kStartErrorStatus, retry_policy, backoff_policy);
+            return make_ready_future(
+                StatusOr<std::unique_ptr<
+                    AsyncStreamingReadWriteRpc<RequestType, ResponseType>>>(
+                    kStartErrorStatus));
+          }
+          return this->initializer_(std::move(potential_stream));
+        });
+
+    start_future_result.then(
+        [this, retry_policy, backoff_policy](
+            future<StatusOr<std::unique_ptr<
+                AsyncStreamingReadWriteRpc<RequestType, ResponseType>>>>
+                start_initialize_future) {
+          auto start_initialize_response = start_initialize_future.get();
+
+          if (!start_initialize_response.ok()) {
+            // Start failure which already invoked retry loop, so return
+            // immediately
+            if (start_initialize_response.status() == kStartErrorStatus) {
+              return;
+            }
+            {
+              std::lock_guard<std::mutex> g{this->mu_};
+              this->in_retry_loop_ = true;
+            }
+            this->AttemptRetry(kInitializerErrorStatus, retry_policy,
+                               backoff_policy);
+            return;
+          }
+
+          std::lock_guard<std::mutex> g{this->mu_};
+          this->stream_ = std::move(*start_initialize_response);
+          if (this->read_promise_.has_value()) {
+            this->read_promise_.value().set_value(absl::optional());
+            this->read_promise_ = absl::optional();
+          }
+          if (this->write_promise_.has_value()) {
+            this->write_promise_.value().set_value(false);
+            this->write_promise_ = absl::optional();
+          }
+          this->in_retry_loop_ = false;
+        });
+  }
+
+  static constexpr kStartErrorStatus = Status(StatusCode.kUnavailable, "Start");
+  static constexpr kReadErrorStatus = Status(StatusCode.kUnavailable, "Read");
+  static constexpr kWriteErrorStatus = Status(StatusCode.kUnavailable, "Write");
+  static constexpr kInitializerErrorStatus =
+      Status(StatusCode.kUnavailable, "Initializer");
+  static constexpr kPermanentErrorStatus =
+      Status(StatusCode.kAborted, "Permanent error");
+
+  std::unique_ptr<RetryPolicy const> const retry_policy_prototype_;
+  std::unique_ptr<BackoffPolicy const> const backoff_policy_prototype_;
+  AsyncSleeper sleeper_;
+  StreamFactory<RequestType, ResponseType> const stream_factory_;
+  StreamInitializer<RequestType, ResponseType> const initializer_;
+
+  std::mutex mu_;
+
+  std::unique_ptr<AsyncStreamingReadWriteRpc<RequestType, ResponseType>>
+      stream_;          // ABSL_GUARDED_BY(mu_)
+  bool in_retry_loop_;  // ABSL_GUARDED_BY(mu_)
+  absl::optional<promise<absl::optional<ResponseType>>>
```

Comment:
Can you make both of these `promise<void>` and call them `read_reinit_done_` and `write_reinit_done_` as well as provide a comment for why they exist and what they do?

Reply:
doesn't the promises need to hold the type of `Read` and `Write` respectively since the futures of these promises are what's being returned in `Read` and `Write`?